tests/sample/morris/test_morris.py

```diff
@@ -13,9 +13,12 @@
                                  compute_b_star,
                                  compute_delta,
                                  generate_trajectory,
-                                 generate_x_star)
+                                 generate_x_star,)
 from SALib.util import read_param_file, compute_groups_matrix
 
+from src.SALib.sample.morris import define_problem_with_groups, \
+    check_if_num_levels_is_even, check_group_membership
+
```

Comment:
Shouldn't need to specify `src` as the top-level module:

```python
from SALib.sample.morris import define_problem_with_groups

# vs

from src.SALib.sample.morris import define_problem_with_groups
```

I prefer using parentheses to group multi-line imports as it looks cleaner than using `\` to me.

```python
from SALib.sample.morris import (define_problem_with_groups,
                                 check_if_num_levels_is_even, 
                                 check_group_membership)
```

Reply:
For some strange reason, when I remove the `src` I get an import error. This is only happening with the functions I created, I don't know exactly why. I tried to google this error, but didn't find a good answer. Can you take a look, please? For now I just left the code as it was.

Reply:
Sure, I will do that asap

Reply:
In `__init__.py` you have to import the high-level functions you wish users to see.

```python
# in __init__.py

from .morris import sample
```

This will allow users to import Morris sampling as before.

```python

from SALib.sample.morris import sample as morris_sample
```

As you want to import specific functions in the context of testing, you have to specify the file itself in the import statement:

```python

# import from the sample dir -> morris dir -> morris file
from SALib.sample.morris.morris import (define_problem_with_groups,
                                        check_if_num_levels_is_even, 
                                        check_group_membership)
```

Alternatively, you could specify those functions in the morris package `__init__.py` as well but this will pollute the namespace and auto-complete, allowing users who don't usually have a reason to be aware of these functions to see them.

Reply:
Hello, @ConnectedSystems. Thanks for your reply. Being relatively new to Python, I'm trying to figure out how exactly all those import statements work. I would like to share what I have seen so far and ask for your opinion (I'm sorry if this reply is too large).

In the module `morris`, inside the subpackage `morris`, module `common_args` is imported using the following command: `from SALib.sample import common_args`.

When I hover my mouse over `SALib`, Pycharm shows me the following warning:

![image](https://user-images.githubusercontent.com/10145845/85335681-a9eee800-b4dd-11ea-965b-39df7cd40cf9.png)

Indeed, this "Package containing module SALib" isn't listed in the requirements,  because that is the package SALib itself. At first I considered this a little strange, because I had to install SALib in my virtual environment to develop SALib. Since all files are in the same project, I believe it shouldn't be necessary to do that, but if I don't, I get an error.

The solution to this would be to import using the folders hierarchy. Since the package `sample` is one level up the package `morris`, this import would become:

`from .. import common_args`

This works and gives no warnings. The same holds to import modules from another package, and this is were this discussion started. 

In my code, inside the `tests` package, I used this: `from src.SALib.sample.morris.morris import (check_group_membership)`. What I'm doing here is to make an absolute import _inside_ the project (as suggested by Pycharm), with no dependencies from any package installed in my virtual env.

In this same package there is another import: `from SALib.sample.morris import (sample)`. This is importing from _outside_ the project, i. e., it's importing from the package SALib installed in my virtual env, not the one I'm currently developing. The downside of this is that, if I make local changes to my modules, this import won't be able to "see" this.

As far as I understand, this is exactly what happened here. I added three extra functions to the package `morris`, but I couldn't import them using just `from SALib.sample.morris.morris` because the installation in my virtual env is now outdated. I would have to install it again, this time with the functions I wrote included (or use another technique to automatically update it, I believe this can be done).

Finally, if everything I wrote is correct, I believe the best option would be:

1. Replace all import statements that use `SALib.<something>` by the correspondent import using the folders hierarchy.
2. Continue to use absolute imports in the package `tests`
3. Expose the method `sample` to the user using the `__init__.py` file, as you suggested

I hope I was clear, and I'm looking forward for your reply. Don't hesitate to ask me if you didn't understand something.

A good reference can be found here: https://realpython.com/lessons/subpackages/

Reply:
Hi,

Not sure I followed everything, I'll have another read through later when my mind is clearer.

One thing that's clear though is that you have multiple copies of SALib floating around.
In the vast majority of cases you shouldn't have to do this.

Set your project up so that you have one (and only one) project folder which you are working on and install the project in "development mode" (or an "editable install"). 

From the project root (top of project directory):

```bash
# Yes the period is important
$ pip install -e .

# Or if using conda and conda-build
$ conda develop .

# Or if you have neither of the above
$ python setup.py develop
```

See here for more explanation:
https://stackoverflow.com/a/61813942/2694952

Doing the above means you don't have to copy files back and forth which is a source of issues when working with many people and branches.

Reply:
Hi, @ConnectedSystems. Thanks for this very interesting info about the develop mode, I wasn't aware about it. 

Actually, I only have one copy of SALib source code. What I was saying is that I installed SALib using regular `pip install salib` directly from PyPI in a virtual environment.

As I understand, there are basically 4 options:

1. Regular `pip install salib` directly from PyPI (as I did). **If I need to have SALib installed to run the tests**, this is a terrible choice, since what I change in the code doesn't affect my installed version, of course. My point is: in my opinion, I shouldn't need to have the package actually installed in my virtual env to run some piece of test code. Since the tests are in the same folder as my source code (but not in the same package), the changes I do should be visible to them.

2. Don't install SALib at all. The way the code is today, this wouldn't work. As I said in my last answer, there are some imports that require SALib to be installed, such as `from SALib.sample import common_args`

3. Install SALib in development mode. This works, and actually fix all the problems. But SALib continues to make imports from itself, what I think it's weird (again, this is just my opinion...)

4. Install SALib in development mode **AND** remove the dependency from SALib itself, as I said in the previous reply. By doing this, SALib won't require itself to be installed to work, and at the same time the new features included would be ready to be used in this virtual env.

Please, let me know what you think about that, I hope I was more clear this time. Nevertheless, I believe it's a little difficult to explain this without showing my screen, so I'm available to a quick video call if you want.

Reply:
Hi again,

A package referencing itself is not that big a deal although I understand your struggle to reconcile the conceptual mismatch. The end result is the same regardless of how the `import` statement is structured as Python resolves the statement to the same file and code.

Given we're developing a package it makes sense for the package to be installed (albeit in development mode).

Regarding the need to have the package installed to run tests: Tests should be run with the package installed. We're not, in this context, testing a loose collection of scripts - we are testing the correct functioning of a package.

Lastly, PEP8 recommends favouring absolute imports with relative imports acceptable when dealing with a complex package layout (https://www.python.org/dev/peps/pep-0008/#imports).

The example given...

```python
import mypkg.sibling
from mypkg import sibling
from mypkg.sibling import example
```

... lines up with how SALib, [and other Python packages](https://github.com/pandas-dev/pandas/blob/master/pandas/core/api.py), approach importing own code.

Hope this clears things up.

Reply:
Hi, @ConnectedSystems.

Thank you for your answer, everything is clear now. I'm happy to be contributing here, I'm learning a lot (I guess this is also the point of contributing to open source projects).

I'll commit my branch asap.